src/Components/Blazor/WebAssembly.Authentication/src/Interop/AuthenticationService.ts

```diff
@@ -0,0 +1,319 @@
+    completeSignOut(url: string): Promise<AuthenticationResult>;
+}
+
+class OidcAuthorizeService implements AuthorizeService {
+    private _userManager: UserManager;
+
+    constructor(userManager: UserManager) {
+        this._userManager = userManager;
+    }
+
+    async getUser() {
+        const user = await this._userManager.getUser();
+        return user && user.profile;
+    }
+
+    async getAccessToken(request?: AccessTokenRequestOptions): Promise<AccessTokenResult> {
+        const user = await this._userManager.getUser();
+        if (hasValidAccessToken(user) && hasAllScopes(request, user.scopes)) {
+            return {
+                status: AccessTokenResultStatus.Success,
+                token: {
+                    grantedScopes: user.scopes,
+                    expires: getExpiration(user.expires_in),
+                    value: user.access_token
+                }
+            };
+        } else {
+            try {
+                const parameters = request && request.scopes ?
+                    { scope: request.scopes.join(' ') } : undefined;
+
+                const newUser = await this._userManager.signinSilent(parameters);
+
+                return {
+                    status: AccessTokenResultStatus.Success,
+                    token: {
+                        grantedScopes: newUser.scopes,
+                        expires: getExpiration(newUser.expires_in),
+                        value: newUser.access_token
+                    }
+                };
+
+            } catch (e) {
+                return {
+                    status: AccessTokenResultStatus.RequiresRedirect
+                };
+            }
+        }
+
+        function hasValidAccessToken(user: User | null): user is User {
+            return !!(user && user.access_token && !user.expired && user.scopes);
+        }
+
+        function getExpiration(expiresIn: number) {
+            const now = new Date();
+            now.setTime(now.getTime() + expiresIn * 1000);
+            return now;
+        }
+
+        function hasAllScopes(request: AccessTokenRequestOptions | undefined, currentScopes: string[]) {
+            const set = new Set(currentScopes);
+            if (request && request.scopes) {
+                for (let current of request.scopes) {
+                    if (!set.has(current)) {
+                        return false;
+                    }
+                }
+            }
+
+            return true;
+        }
+    }
+
+    async signIn(state: any) {
+        try {
+            await this._userManager.clearStaleState();
+            const silentUser = await this._userManager.signinSilent(this.createArguments());
+            return this.success(state);
+        } catch (silentError) {
+            await this._userManager.clearStaleState();
+            // User might not be authenticated, fallback to popup authentication
+            console.log("Silent authentication error: ", silentError);
```

Comment:
Nit: random variation in quote types (sorry for even pointing this out)

Reply:
More importantly, why are we logging this stuff to the console directly? Shouldn't the log messages get into .NET where they can go through your `ILogger`?

If you feel this is not likely to happen often and is just an implementation detail we could change later, that's OK with me, since I know it's really incovenient to be redirecting the logs asynchronously via .NET.

On the other hand if there are realistically going to be cases where these log messages are *important* and actually must go through the configured `ILogger`, let's consider alternatives. For example since this is specific to WebAssembly, we could consider adding a synchronous interop call to send a log message back to .NET from here.

Reply:
This was done.